docs: Add dependency injection tutorial (#424)
unoplatform/uno.extensions: doc/How-Tos/DependencyInjection/DependencyInjection-setup.md

## Step-by-steps

1. Plan the contract for your service
    * Create a new interface which declares the method(s) your service offers:
    ```cs
        public interface IUserProfilePictureService
        {
	        Task<byte[]> GetAsync(CancellationToken ct = null);
        }
    ```

2. Add & register a service that encapsulates app functionality
    * Write a new service class which implements the interface you created above, defining its member(s):
    ```cs
        public class UserProfilePictureService : IUserProfilePictureService
        {
            public async Task<byte[]> GetAsync(CancellationToken ct = null)
            {
                . . .
            }
        }
    ```
    * Register this service implementation with the `IServiceProvider` instance provided by your application's `IHostBuilder`:
    ```cs
        public IHost Host { get; init; }

        public App()
        {
            Host = UnoHost
                .CreateDefaultBuilder()
                .ConfigureServices(services =>
				{
					// Register your services below
					services.AddSingleton<IUserProfilePictureService, UserProfilePictureService>();
				})
                .Build();
        }
    ```

> 3. Get & use an instance of the registered service

Reviewer: Change to "Get and use" rather than & (same goes for elsewhere in this doc)

    * You can now retrieve this service in its instantiated form from anywhere in the application such as a view model:
    ```cs
        public class MainViewModel
        {
            private IUserProfilePictureService userPhotoService;

            public MainViewModel()
            {

>                 userPhotoService = (Application.Current as App).Host.Services.GetService<IUserProfilePictureService>();

Reviewer: Use constructor injection here - we should avoid using GetService wtihin a viewmodel
You need to register both service and viewmodel and then in codebehind for mainPage call GetRequiredService<MainViewModel>

# How-To: Use Services with Dependency Injection

> Dependency injection (DI) is an important design pattern for building loosely-coupled software that allows for maintainability and testing. This tutorial will walk you through how to register services with a service provider for use throughout your application.

Reviewer: How do we feel about linking out to other sites for more detailed explanation on DI and design patterns
cc @francoistanguay @jeromelaban

Author: This is still a pending question, as there are some helpful resources we could add subtle references to. Notably, these:
https://docs.microsoft.com/en-us/dotnet/core/extensions/dependency-injection
https://docs.microsoft.com/en-us/dotnet/architecture/modern-web-apps-azure/architectural-principles#dependency-inversion

Thoughts @jeromelaban @francoistanguay?

> NOTE: This guide assumes you used the Uno.Extensions `dotnet new` template to create the solution. You can verify this by looking for a generated `App.xaml.host.cs` file.

Reviewer 2: ```suggestion
> [!Tip] This guide assumes you used the Uno.Extensions `dotnet new unoapp-extensions` template to create the solution. You can verify this by looking for a generated `App.xaml.host.cs` file.
```

When you say generated, does it mean that it was generated when creating the template?

Reviewer: Yeh, the app.xaml.host.cs file is generated by the template. Change "for a generated" to "for the generated"
Also, suggest we link to the getting started doc here in case they need to walk through creating the app from the template

## Step-by-steps

1. Plan the contract for your service
    * Create a new interface which declares the method(s) your service offers:
    ```cs
        public interface IUserProfilePictureService
        {

> 	        Task<byte[]> GetAsync(CancellationToken ct = null);

Reviewer 2: Should we already imply nullables to be enabled?
```suggestion
	        Task<byte[]> GetAsync(CancellationToken? ct = null);
```

Reviewer: Actually I think it would be better practice to remove the default completely here
Task<byte[]> GetAsync(CancellationToken ct);

Also might be worth simplifying the name of this service to just IProfilePictureService

        }
    ```

2. Add and register a service that encapsulates app functionality
    * Write a new service class which implements the interface you created above, defining its member(s):
    ```cs
        public class UserProfilePictureService : IUserProfilePictureService
        {

>             public async Task<byte[]> GetAsync(CancellationToken ct = null)

Reviewer 2: ```suggestion
            public async Task<byte[]> GetAsync(CancellationToken? ct = null)
```

            {
                . . .
            }
        }
    ```

>     * Register this service implementation with the `IServiceProvider` instance provided by your application's `IHostBuilder`:

Reviewer: This isn't correct - you register on the IServiceCollection and then resolve on the IServiceProvider instance

                .ConfigureServices(services =>
				{
					// Register your services below
					services.AddSingleton<IUserProfilePictureService, UserProfilePictureService>();
				})
                .Build();
        }
    ```
3. Leverage constructor injection to use an instance of the registered service
    * Create a new view model class that will use the functionality offered by your service. Add a constructor with a parameter of the same type as the service interface you defined earlier:
    ```cs
        public class MainViewModel
        {
            private readonly IUserProfilePictureService userPhotoService;

            public MainViewModel(IUserProfilePictureService userPhotoService)
            {
                this.userPhotoService = userPhotoService;
            }
        }
    ```
    * For the dependency injection framework to handle instantiation of the service as a constructor argument, you must also register your view model with the IServiceProvider:
    ```cs
        public IHost Host { get; private set; }

        public App()
        {
            Host = UnoHost
                .CreateDefaultBuilder()
                .ConfigureServices(services =>
				{

> 					// Register your services below

Reviewer: Trim "below"

					services.AddSingleton<IUserProfilePictureService, UserProfilePictureService>();
                    // Register view model below
                    services.AddTransient<MainViewModel>();
				})
                .Build();
        }        
    ```
    * Now, `MainViewModel` has access to the functionality provided by the resolved implementation of your service:
    ```cs
        byte[] profilePhotoBytes = await userPhotoService.GetAsync();
    ```
4. Associate the view model with a view from code behind

>     * From the code behind of a view, directly reference the application's IHost instance to request an instance of the desired view model. Set this as the DataContext:

Reviewer 2: ```suggestion
    * From the code behind of a view, directly reference the application's `IHost` instance to request an instance of the desired view model. Set this as the `DataContext`:
```
